x/stake/handler.go

```diff
@@ -53,212 +63,245 @@ func EndBlocker(ctx sdk.Context, k Keeper) (ValidatorUpdates []abci.Validator) {
 
-	// test that there are enough shares to unbond
-	if msg.Shares == "MAX" {
-		if !bond.Shares.GT(sdk.ZeroRat()) {
-			return ErrNotEnoughBondShares(k.codespace, msg.Shares).Result()
-		}
-	} else {
-		var err sdk.Error
-		delShares, err = sdk.NewRatFromDecimal(msg.Shares)
+	k.CoinKeeper().AddCoins(ctx, ubd.DelegatorAddr, sdk.Coins{ubd.Balance})
+	k.RemoveUnbondingDelegation(ctx, ubd)
+
+	tags := sdk.NewTags(
+		TagAction, ActionCompleteUnbonding,
+		TagDelegator, msg.DelegatorAddr.Bytes(),
+		TagSrcValidator, msg.ValidatorAddr.Bytes(),
+	)
+
+	// add slashed tag only if there has been some slashing
+	if !ubd.Slashed.IsZero() {
+		bz, err := json.Marshal(ubd.Slashed)
 		if err != nil {
-			return err.Result()
-		}
-		if bond.Shares.LT(delShares) {
-			return ErrNotEnoughBondShares(k.codespace, msg.Shares).Result()
+			panic(err)
 		}
+		tags = tags.AppendTag(string(TagSlashed), bz)
 	}
+	return sdk.Result{Tags: tags}
+}
 
-	// get validator
-	validator, found := k.GetValidator(ctx, msg.ValidatorAddr)
-	if !found {
-		return ErrNoValidatorForAddress(k.codespace).Result()
-	}
+func handleMsgBeginRedelegate(ctx sdk.Context, msg types.MsgBeginRedelegate, k keeper.Keeper) sdk.Result {
 
-	if ctx.IsCheckTx() {
-		return sdk.Result{}
+	delegation, srcValidator, pool, returnAmount, err := k.UnbondDelegation(ctx, msg.DelegatorAddr, msg.ValidatorSrcAddr, msg.SharesAmount)
+	if err != nil {
+		return err.Result()
 	}
 
-	// retrieve the amount of bonds to remove (TODO remove redundancy already serialized)
-	if msg.Shares == "MAX" {
-		delShares = bond.Shares
+	// update then remove the source validator if necessary
+	srcValidator = k.UpdateValidator(ctx, srcValidator)
+	if srcValidator.DelegatorShares.IsZero() {
+		k.RemoveValidator(ctx, srcValidator.Owner)
 	}
 
-	// subtract bond tokens from delegator bond
-	bond.Shares = bond.Shares.Sub(delShares)
+	params := k.GetParams(ctx)
+	returnCoin := sdk.Coin{params.BondDenom, sdk.NewInt(returnAmount)}
+	dstValidator, found := k.GetValidator(ctx, msg.ValidatorSrcAddr)
+	if !found {
+		return ErrBadRedelegationDst(k.Codespace()).Result()
+	}
+	sharesCreated, delegation, dstValidator, pool, err := k.Delegate(ctx, msg.DelegatorAddr, returnCoin, dstValidator)
+	k.SetPool(ctx, pool)
+	k.SetDelegation(ctx, delegation)
+	k.UpdateValidator(ctx, dstValidator)
+
+	// create the unbonding delegation
+	minTime := ctx.BlockHeader().Time + params.UnbondingTime
+	minHeight := ctx.BlockHeight() + params.MinUnbondingBlocks
+
+	red := Redelegation{
+		DelegatorAddr:    msg.DelegatorAddr,
+		ValidatorSrcAddr: msg.ValidatorSrcAddr,
+		ValidatorDstAddr: msg.ValidatorDstAddr,
+		MinTime:          minTime,
+		MinHeight:        minHeight,
+		SharesDst:        sharesCreated,
+		SharesSrc:        msg.SharesAmount,
+	}
+	k.SetRedelegation(ctx, red)
 
-	// remove the bond
-	revokeValidator := false
-	if bond.Shares.IsZero() {
+	tags := sdk.NewTags(
+		tags.Action, tags.ActionBeginRedelegation,
+		tags.Delegator, msg.DelegatorAddr.Bytes(),
```

Comment:
Ditto as above (address encoding in tags) for these three lines

Reply:
addressed